### PR TITLE
Move debug to safe native + implement dummy to make old framework compatible

### PR DESCRIPTION
**Pull request by gerben-stavenga**
Solve the incompatibility issue with new binary + old framework

### <samp>🤖 Generated by Copilot at c803a49</samp>

The pull request updates the debug module to use the new safe native function framework, which improves safety and flexibility of native functions. It also adds a new native_format_debug function to the string_utils module, which is used by the debug module to format values for debug output. The changes are compatible with the old debug functions for replayability on-chain.

### <samp>🤖 Generated by Copilot at c803a49</samp>

*  Update imports, signatures, and implementations of native debug functions to use the new safe native function framework ([link](https://github.com/aptos-labs/aptos-core/pull/8017/files?diff=unified&w=0#diff-d0741d3f74fd704d43cafab775e97054a6e357541e7815621d4bce648d52c89cL7-R25), [link](https://github.com/aptos-labs/aptos-core/pull/8017/files?diff=unified&w=0#diff-d0741d3f74fd704d43cafab775e97054a6e357541e7815621d4bce648d52c89cL25-R43), [link](https://github.com/aptos-labs/aptos-core/pull/8017/files?diff=unified&w=0#diff-d0741d3f74fd704d43cafab775e97054a6e357541e7815621d4bce648d52c89cL39-R54), [link](https://github.com/aptos-labs/aptos-core/pull/8017/files?diff=unified&w=0#diff-d0741d3f74fd704d43cafab775e97054a6e357541e7815621d4bce648d52c89cL57-R66), [link](https://github.com/aptos-labs/aptos-core/pull/8017/files?diff=unified&w=0#diff-d0741d3f74fd704d43cafab775e97054a6e357541e7815621d4bce648d52c89cL71-R114))
*  Remove make_native_print and make_native_stack_trace functions, and replace them with make_safe_native macro calls that wrap the native functions and handle common logic and error handling ([link](https://github.com/aptos-labs/aptos-core/pull/8017/files?diff=unified&w=0#diff-d0741d3f74fd704d43cafab775e97054a6e357541e7815621d4bce648d52c89cL39-R54), [link](https://github.com/aptos-labs/aptos-core/pull/8017/files?diff=unified&w=0#diff-d0741d3f74fd704d43cafab775e97054a6e357541e7815621d4bce648d52c89cL71-R114))
*  Add native_old_debug_print and native_old_print_stacktrace functions to provide dummy implementations of the old debug functions for replayability on-chain, and check for testing feature flag to control debug output ([link](https://github.com/aptos-labs/aptos-core/pull/8017/files?diff=unified&w=0#diff-d0741d3f74fd704d43cafab775e97054a6e357541e7815621d4bce648d52c89cL71-R114))
*  Make native_format_debug function public and crate-visible in `string_utils.rs`, and use it to format values for debug output in the old debug functions ([link](https://github.com/aptos-labs/aptos-core/pull/8017/files?diff=unified&w=0#diff-98b1b9150f61f127f6a7e4c554a4866e149a716b5661ce2fffe7964110d1d093R292-R315), [link](https://github.com/aptos-labs/aptos-core/pull/8017/files?diff=unified&w=0#diff-d0741d3f74fd704d43cafab775e97054a6e357541e7815621d4bce648d52c89cL71-R114))
*  Pass timed features and features parameters to the make_all function in the debug module, and to the make_safe_native macro calls for each native function ([link](https://github.com/aptos-labs/aptos-core/pull/8017/files?diff=unified&w=0#diff-d0741d3f74fd704d43cafab775e97054a6e357541e7815621d4bce648d52c89cL39-R54), [link](https://github.com/aptos-labs/aptos-core/pull/8017/files?diff=unified&w=0#diff-d0741d3f74fd704d43cafab775e97054a6e357541e7815621d4bce648d52c89cL71-R114), [link](https://github.com/aptos-labs/aptos-core/pull/8017/files?diff=unified&w=0#diff-d0741d3f74fd704d43cafab775e97054a6e357541e7815621d4bce648d52c89cL93-R151), [link](https://github.com/aptos-labs/aptos-core/pull/8017/files?diff=unified&w=0#diff-c4be656b5e1494c79e37e28860af5cecf5c3351dd2541397ff6776cfe126177aL432-R436))
*  Move format_debug function from `string_utils.rs` to `debug.rs`, and rename it to native_format_debug ([link](https://github.com/aptos-labs/aptos-core/pull/8017/files?diff=unified&w=0#diff-98b1b9150f61f127f6a7e4c554a4866e149a716b5661ce2fffe7964110d1d093R292-R315), [link](https://github.com/aptos-labs/aptos-core/pull/8017/files?diff=unified&w=0#diff-d0741d3f74fd704d43cafab775e97054a6e357541e7815621d4bce648d52c89cL71-R114))

### CODE REVIEW

File: aptos-move/framework/src/natives/debug.rs [hunk 1]
```diff
@@ -68,35 +74,81 @@ fn native_stack_trace(
...
+                (),
+                timed_features.clone(),
+                features.clone(),
+                native_old_debug_print,
+            ),
+        ),
+        (
+            "print_stack_trace",
+            make_safe_native((), timed_features, features, native_old_print_stacktrace),
+        ),
```
Comment (gregnazario) [inline aptos-move/framework/src/natives/debug.rs:151]: Are these supposed to still be dummies?

**Review by gregnazario — APPROVED** [review]
I've verified that it works, let's just ensure that it doesn't run when onchain.

File: aptos-move/framework/src/natives/debug.rs (hunk 1 above)
Comment (gerben-stavenga) [reply aptos-move/framework/src/natives/debug.rs]: Updated the comment

**Comment by gerben-stavenga** [comment]
> 

What do you mean it doesnt run on-chain? It will run on-chain just not print anything due to "if cfg!(feature == "testing")" part. Just like the before my changes

**Review by davidiw — APPROVED** [review]

**Pull request merged**